### PR TITLE
HIVE-15112: Implement Parquet vectorization reader for Struct type

### PR DESCRIPTION
**Pull request by winningsix**
Refactor UT

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetColumnReader.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.apache.hadoop.hive.ql.io.parquet.vector;
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetColumnReader.java]: License header?

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetColumnReader.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+import org.apache.hadoop.hive.ql.exec.vector.ColumnVector;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+
+import java.io.IOException;
+
+public interface VectorizedParquetColumnReader {
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetColumnReader.java]: Add some comment on the `readBatch` method? Also, from the method signature it seems it should not be restricted to only Parquet. How about `VectorizedColumnReader`?

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetMapReader.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package org.apache.hadoop.hive.ql.io.parquet.vector;
+
+import org.apache.hadoop.hive.ql.exec.vector.ColumnVector;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+
+import java.io.IOException;
+
+public class VectorizedParquetMapReader implements VectorizedParquetColumnReader{
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetMapReader.java]: Is this really necessary? this is the same as `VectorizedParquetColumnReader`.

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetMapReader.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+  @Override
+  public void readBatch(
+    int total,
+    ColumnVector column,
+    TypeInfo columnType) throws IOException {
+
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetMapReader.java]: Extra line

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java [hunk 5]
```diff
@@ -278,12 +284,77 @@ private void checkEndOfRowGroup() throws IOException {
...
     totalCountLoadedSoFar += pages.getRowCount();
   }
+
+  private List<ColumnDescriptor> getAllColumnDescriptorByType(
+    int depth,
+    Type type,
+    List<ColumnDescriptor> columns) {
+    List<ColumnDescriptor> res = new ArrayList<>();
+    for (ColumnDescriptor descriptor : columns) {
+      if (type.getName().equals(descriptor.getPath()[depth])) {
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java]: What if the path length is smaller than `depth`? Will this crash?

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java [hunk 6]
```diff
@@ -278,12 +284,77 @@ private void checkEndOfRowGroup() throws IOException {
...
+      List<Type> types = type.asGroupType().getFields();
+      for (int i = 0; i < fieldTypes.size(); i++) {
+        VectorizedParquetColumnReader r =
+          buildVectorizedParquetReader(fieldTypes.get(i), types.get(i), pages, descriptors,
+            skipTimestampConversion, depth + 1);
+        if (r != null) {
+          fieldReaders.add(r);
+        }
+      }
+      if (fieldReaders.size() > 0) {
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java]: What if `fieldReaders.size()` is not equal to `fieldTypes.size()`. Can this be handled?

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedStructReader.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
+package org.apache.hadoop.hive.ql.io.parquet.vector;
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedStructReader.java]: License header

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedPrimitiveColumnReader.java [hunk 8]
```diff
@@ -155,48 +159,96 @@ void readBatch(
...
+    int num,
+    ColumnVector column,
+    TypeInfo columnType,
+    int rowId) throws IOException {
+    switch (columnType.getCategory()) {
+    case PRIMITIVE:
+      PrimitiveTypeInfo primitiveColumnType = (PrimitiveTypeInfo) columnType;
+      readBatchForPrimitiveType(num, column, primitiveColumnType, rowId);
+      break;
+    case LIST:
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedPrimitiveColumnReader.java]: If this is a primitive column reader, why it should read complex types?

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java (hunk 5 above)
Comment (winningsix) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java]: It happens only when schema is corrupted. Addressing this by adding a check before this if block.

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java [hunk 9]
```diff
@@ -279,11 +283,82 @@ private void checkEndOfRowGroup() throws IOException {
...
     totalCountLoadedSoFar += pages.getRowCount();
   }
+
+  private List<ColumnDescriptor> getAllColumnDescriptorByType(
+    int depth,
+    Type type,
+    List<ColumnDescriptor> columns) throws ParquetRuntimeException {
+    List<ColumnDescriptor> res = new ArrayList<>();
+    for (ColumnDescriptor descriptor : columns) {
+      if (depth > descriptor.getPath().length) {
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java]: `>=` ?

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java [hunk 10]
```diff
@@ -279,11 +283,82 @@ private void checkEndOfRowGroup() throws IOException {
...
+
+  // Build VectorizedParquetColumnReader via Hive typeInfo and Parquet schema
+  private VectorizedColumnReader buildVectorizedParquetReader(
+    TypeInfo typeInfo,
+    Type type,
+    PageReadStore pages,
+    MessageType schema,
+    boolean skipTimestampConversion) throws IOException {
+    return buildVectorizedParquetReader(typeInfo, type, pages, schema.getColumns(), skipTimestampConversion,
+      0);
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java]: nit: can we put this into the same line above? for easier reading.

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java [hunk 11]
```diff
@@ -279,11 +283,82 @@ private void checkEndOfRowGroup() throws IOException {
     List<ColumnDescriptor> columns = requestedSchema.getColumns();
     List<Type> types = requestedSchema.getFields();
     columnReaders = new VectorizedColumnReader[columns.size()];
-    for (int i = 0; i < columns.size(); ++i) {
+    for (int i = 0; i < types.size(); ++i) {
       columnReaders[i] =
-        new VectorizedColumnReader(columns.get(i), pages.getPageReader(columns.get(i)),
-          skipTimestampConversion, types.get(i));
+        buildVectorizedParquetReader(columnTypesList.get(indexColumnsWanted.get(i)), types.get(i),
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedParquetRecordReader.java]: is it possible that `indexColumnsWanted` could be empty?

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedStructReader.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+import org.apache.hadoop.hive.ql.exec.vector.StructColumnVector;
+import org.apache.hadoop.hive.serde2.typeinfo.StructTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+
+import java.io.IOException;
+import java.util.List;
+
+public class VectorizedStructReader implements VectorizedColumnReader {
+
+  List<VectorizedColumnReader> fieldReaders;
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedStructReader.java]: make this private?

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedStructReader.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    ColumnVector column,
+    TypeInfo columnType) throws IOException {
+    StructColumnVector structColumnVector = (StructColumnVector) column;
+    StructTypeInfo structTypeInfo = (StructTypeInfo) columnType;
+    ColumnVector[] vectors = structColumnVector.fields;
+    for (int i = 0; i < vectors.length; i++) {
+      fieldReaders.get(i)
+        .readBatch(total, vectors[i], structTypeInfo.getAllStructFieldTypeInfos().get(i));
+      structColumnVector.isRepeating = structColumnVector.isRepeating && vectors[i].isRepeating;
+      for (int j = 0; j < vectors[i].isNull.length; j++) {
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedStructReader.java]: I think there's a difference between null struct versus struct with null fields. Seems this treat the two cases as the same. Do we need to differentiate them?

File: ql/src/test/org/apache/hadoop/hive/ql/io/parquet/TestVectorizedColumnReader.java [hunk 14]
```diff
@@ -14,416 +18,64 @@
...
-      + "} ");
-
-  @AfterClass
-  public static void cleanup() throws IOException {
-    FileSystem fs = file.getFileSystem(conf);
-    if (fs.exists(file)) {
-      fs.delete(file, true);
-    }
-  }
+public class TestVectorizedColumnReader extends TestVectorizedColumnReaderBase{
```
Comment (sunchao) [inline ql/src/test/org/apache/hadoop/hive/ql/io/parquet/TestVectorizedColumnReader.java]: space before {

File: ql/src/test/org/apache/hadoop/hive/ql/io/parquet/TestVectorizedColumnReaderBase.java [hunk 15]
```diff
@@ -0,0 +1,594 @@
...
+        }
+        assertEquals("No Null check failed at " + c, noNull, vector.noNulls);
+        assertFalse(vector.isRepeating);
+      }
+      assertEquals("It doesn't exit at expected position", nElements, c);
+    } finally {
+      reader.close();
+    }
+  }
+
```
Comment (sunchao) [inline ql/src/test/org/apache/hadoop/hive/ql/io/parquet/TestVectorizedColumnReaderBase.java]: I think we talked about testing reading decimal. Should we add in this patch?

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedStructReader.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+import org.apache.hadoop.hive.ql.exec.vector.StructColumnVector;
+import org.apache.hadoop.hive.serde2.typeinfo.StructTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+
+import java.io.IOException;
+import java.util.List;
+
+public class VectorizedStructReader implements VectorizedColumnReader {
+
+  private List<VectorizedColumnReader> fieldReaders;
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedStructReader.java]: nit: mark this as final?

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedStructReader.java [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    TypeInfo columnType) throws IOException {
+    StructColumnVector structColumnVector = (StructColumnVector) column;
+    StructTypeInfo structTypeInfo = (StructTypeInfo) columnType;
+    ColumnVector[] vectors = structColumnVector.fields;
+    for (int i = 0; i < vectors.length; i++) {
+      fieldReaders.get(i)
+        .readBatch(total, vectors[i], structTypeInfo.getAllStructFieldTypeInfos().get(i));
+      structColumnVector.isRepeating = structColumnVector.isRepeating && vectors[i].isRepeating;
+
+      for (int j = 0; j < vectors[i].isNull.length; j++) {
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedStructReader.java]: should we set `structColumnVector.noNulls` as well?

File: ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedStructColumnReader.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+    ColumnVector[] vectors = structColumnVector.fields;
+    for (int i = 0; i < vectors.length; i++) {
+      fieldReaders.get(i)
+        .readBatch(total, vectors[i], structTypeInfo.getAllStructFieldTypeInfos().get(i));
+      structColumnVector.isRepeating = structColumnVector.isRepeating && vectors[i].isRepeating;
+
+      for (int j = 0; j < vectors[i].isNull.length; j++) {
+        structColumnVector.isNull[j] =
+          (i == 0) ? vectors[i].isNull[j] : structColumnVector.isNull[j] && vectors[i].isNull[j];
+        structColumnVector.noNulls = (i == 0) ? structColumnVector.isNull[j] :
```
Comment (sunchao) [inline ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedStructColumnReader.java]: Hmm.. why this needs to be in the inner loop. Can you just do:
`structColumnVector.noNulls = (i == 0) ? vectors[i].noNulls : structColumnVector.noNulls && vectors[i].noNulls;`?

**Pull request closed without merging**